pypesto/objective/history.py

```diff
@@ -331,128 +530,109 @@ def init_trace(self, x: np.ndarray):
 
-    def extract_values(self, sensi_orders, mode, result):
-        """Extract values to record from result."""
-        if mode == MODE_FUN:
-            fval = np.NaN if 0 not in sensi_orders \
-                else result.get(FVAL, np.NaN)
-            grad = None if not self.options.trace_record_grad \
-                or 1 not in sensi_orders \
-                else result.get(GRAD, None)
-            hess = None if not self.options.trace_record_hess \
-                or 2 not in sensi_orders \
-                else result.get(HESS, None)
-            res = None
-            sres = None
-            chi2 = np.NaN
-            schi2 = None
-        else:  # mode == MODE_RES
-            res_result = result.get(RES, None)
-            sres_result = result.get(SRES, None)
-            chi2 = np.NaN if not self.options.trace_record_chi2 \
-                or 0 not in sensi_orders \
-                else res_to_chi2(res_result)
-            schi2 = None if not self.options.trace_record_schi2 \
-                or 1 not in sensi_orders \
-                else sres_to_schi2(res_result, sres_result)
-            fval = np.NaN if 0 not in sensi_orders \
-                else result.get(FVAL, chi2)
-            grad = None if not self.options.trace_record_grad \
-                or 1 not in sensi_orders \
-                else schi2
-            hess = None if not self.options.trace_record_hess \
-                or 1 not in sensi_orders \
-                else sres_to_fim(sres_result)
-            res = None if not self.options.trace_record_res \
-                or 0 not in sensi_orders \
-                else res_result
-            sres = None if not self.options.trace_record_sres \
-                or 1 not in sensi_orders \
-                else sres_result
-
-        return fval, grad, hess, res, sres, chi2, schi2
-
-    def save_trace(self, finalize: bool = False):
+    def _save_trace(self, finalize: bool = False):
         """
         Save to file via pd.DataFrame.to_csv() if `self.storage_file` is
         not None and other conditions apply.
 
         .. note::
             Format might be revised when storage is implemented.
         """
-        if self.storage_file is None:
+        if self.file is None:
             return
 
         if finalize \
-                or (len(self.trace) > 0 and len(self.trace) %
+                or (len(self._trace) > 0 and len(self._trace) %
                     self.options.trace_save_iter == 0):
             # save
-            trace_copy = copy.deepcopy(self.trace)
+            trace_copy = copy.deepcopy(self._trace)
             for field in [('hess', np.NaN), ('res', np.NaN), ('sres', np.NaN)]:
                 trace_copy[field] = trace_copy[field].apply(
                     ndarray2string_full
                 )
-            trace_copy.to_csv(self.storage_file)
+            trace_copy.to_csv(self.file)
+
+    def get_fval_trace(self) -> Sequence[float]:
```

Comment:
pretty sure the rtype here is incorrect, see above

Reply:
I will check if python allows a more precise return type in overriding functions. if yes, I will adjust.